### PR TITLE
2296 add otelcol processor tail sampling component

### LINKED ISSUE #2296: Flow: otelcol.processor.tail_sampling component
Expose the OpenTelemtry Collector `tail_sampling` processor as a Flow component called `otelcol.processor.tail_sampling`. This component is from the `otelcol-contrib` distribution.

### PR DESCRIPTION
**Pull request by erikbaranowski**
#### PR Description

implements a flow component otelcol.processor.tail_sampling that wraps over the upstream OpenTelemetry Collector Contrib `tail_sampling` processor.

#### Which issue(s) this PR fixes

Closes #2296 

#### Notes to the Reviewer

This has a rather lengthy component config and doc.

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -168,7 +168,7 @@ lint: agentlint
 # more without -race for packages that have known race detection issues.
 test:
 	$(GO_ENV) go test $(GO_FLAGS) -race ./...
-	$(GO_ENV) go test $(GO_FLAGS) ./pkg/integrations/node_exporter ./pkg/logs ./pkg/operator ./pkg/util/k8s
+	$(GO_ENV) go test $(GO_FLAGS) ./pkg/integrations/node_exporter ./pkg/logs ./pkg/operator ./pkg/util/k8s ./component/otelcol/processor/tail_sampling
```
Comment (erikbaranowski) [inline Makefile:171]: There is a DATA RACE in the downstream code for this package

**Review by rfratto — APPROVED** [review]
LGTM! Once @karengermond is ok with this and you addressed all comments I'll merge. Thanks!

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+
+## Debug information
+
+`otelcol.processor.tail_sampling` does not expose any component-specific debug
+information.
+
+## Example
+
+This example batches trace data from the agent before sending it to
+[otelcol.exporter.otlp][] for further processing. This example shows an impracticle amount of policies for the purpose of demonstrating how to set each type up:
```
Comment (karengermond) [inline docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md:299]:
```suggestion
This example batches trace data from Grafana Agent before sending it to
[otelcol.exporter.otlp][] for further processing. This example shows an impractical number of policies for the purpose of demonstrating how to set up each type.
```

**Review by karengermond — APPROVED** [review]
LGTM! minor updates

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+### string_attribute block
+
+The `string_attribute` block configures a policy of type `string_attribute`. The policy will sample based on string attributes (resource and record) value matches, both exact and regex value matches are supported.
+
+The following arguments are supported:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`key`                    | `string`    | Tag that the filter is going to be matching against. | | yes
+`values`                 | `string` [] | Values indicate the set of values or regular expressions to use when matching against attribute values. | | yes
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md]: all set (had conflicting suggestion on line so I manually implemented)

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+### trace_state block
+
+The `trace_state` block configures a policy of type `trace_state`. The policy will sample based on TraceState value matches.
+
+The following arguments are supported:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`key`                    | `string`    | Tag that the filter is going to be matching against. | | yes
+`values`                 | `string` [] | Values indicate the set of values to use when matching against trace_state values. | | yes
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md]: all set (had conflicting suggestion on line so I manually implemented)

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+
+`otelcol.processor.tail_sampling` does not expose any component-specific debug
+information.
+
+## Example
+
+This example batches trace data from the agent before sending it to
+[otelcol.exporter.otlp][] for further processing. This example shows an impracticle amount of policies for the purpose of demonstrating how to set each type up:
+
+```river
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md:283]: all set

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md]: all set

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+  }
+  policy {
+    ...
+  }
+  ...
+
+  output {
+    traces  = [...]
+  }
+}
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md]: all set

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+policy > composite > composite_sub_policy > probabilistic     | [probabilistic] | The policy will sample a percentage of traces. | no
+policy > composite > composite_sub_policy > status_code       | [status_code] | The policy will sample based upon the status code. | no
+policy > composite > composite_sub_policy > string_attribute  | [string_attribute] | The policy will sample based on string attributes (resource and record) value matches. | no
+policy > composite > composite_sub_policy > rate_limiting     | [rate_limiting] | The policy will sample based on rate. | no
+policy > composite > composite_sub_policy > span_count        | [span_count] | The policy will sample based on the minimum number of spans within a batch. | no
+policy > composite > composite_sub_policy > trace_state       | [trace_state] | The policy will sample based on TraceState value matches. | no
+output                                                        | [output] [] | Configures where to send received telemetry data. | yes
+
+[policy]: #policy-block
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md:89]: all set

File: component/otelcol/processor/tail_sampling/tail_sampling_test.go [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+	"github.com/grafana/agent/pkg/river"
+	"github.com/grafana/agent/pkg/util"
+	"github.com/grafana/dskit/backoff"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+func TestBadRiverConfig(t *testing.T) {
+	var args Arguments
+	require.Error(t, river.Unmarshal([]byte(exampleBadRiverConfig), &args), "num_traces must be greater than zero")
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply component/otelcol/processor/tail_sampling/tail_sampling_test.go:37]: All set

File: component/otelcol/processor/tail_sampling/types.go [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+		"string_attribute":  andSubPolicyCfg.StringAttributeCfg.Convert(),
+		"rate_limiting":     andSubPolicyCfg.RateLimitingCfg.Convert(),
+		"span_count":        andSubPolicyCfg.SpanCountCfg.Convert(),
+		"trace_state":       andSubPolicyCfg.TraceStateCfg.Convert(),
+	}, &otelCfg)
+
+	return otelCfg
+}
+
+func decodeMapStructure(source map[string]interface{}, otelCfg interface{}) {
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply component/otelcol/processor/tail_sampling/types.go]: All set

File: component/otelcol/processor/tail_sampling/tail_sampling.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+	return nil
+}
+
+// Convert implements processor.Arguments.
+func (args Arguments) Convert() otelconfig.Processor {
+	// TODO: Get rid of mapstructure once tailsamplingprocessor.Config has all public types
+	var otelConfig tsp.Config
+
+	otelPolicyCfgs := []tsp.PolicyCfg{}
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply component/otelcol/processor/tail_sampling/tail_sampling.go]: All set, I changed a few other similar places.

File: component/otelcol/processor/tail_sampling/tail_sampling.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	for _, policyCfg := range args.PolicyCfgs {
+		otelPolicyCfgs = append(otelPolicyCfgs, policyCfg.Convert())
+	}
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply component/otelcol/processor/tail_sampling/tail_sampling.go:81]: I'm not sure I follow on this 1. Are you thinking to decode the parent before the children in the mapstructure?

[the remark this answers is not in the document]
Comment (rfratto) [reply component/otelcol/processor/tail_sampling/tail_sampling.go:81]: Whoops, I misread this and the other ones. It's fine as-is, sorry!

File: docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+
+## Arguments
+
+`otelcol.processor.tail_sampling` supports the following arguments:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`decision_wait`               | `duration` | Wait time since the first span of a trace before making a sampling decision. | `"30s"` | no
+`num_traces`                  | `int`      | Number of traces kept in memory. | `50000` | no
+`expected_new_traces_per_sec` | `int`      | Expected number of new traces (helps in allocating data structures). | `0` | no
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/otelcol.processor.tail_sampling.md:49]: attempted below, let me know if you are looking for different.

**Review by karengermond — APPROVED** [review]
LGTM!

**Pull request merged**